File: 23-merge-k-sorted-lists/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        if not lists:
            return None

        n = len(lists)
        
        ans = []
        for i in range(n):
            
            if not lists[i]:
                continue

            head = lists[i]
            while head is not None:
                ans.append(head.val)
                head = head.next

        if not ans:
            return None

        ans.sort()

        dummy = ListNode()
        current = dummy

        for val in ans:
            current.next = ListNode(val)
            current = current.next
        
        return dummy.next
        

File: 23-merge-k-sorted-lists/main.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
    
    if lists == nil {
        return nil
    }

    ans := []int{}
    for _, list := range lists {
        head := list
        for head != nil {
            ans = append(ans, head.Val)
            head = head.Next
        }
    }

    if len(ans) == 0 {
        return nil
    }

    sort.Slice(ans, func(i, j int) bool {
        return ans[i] < ans[j]
    })

    dummy := &ListNode{Val: 0, Next: nil}
    current := dummy

    for _, val := range ans {
        current.Next = &ListNode{Val: val}
        current = current.Next
    }

    return dummy.Next
}
